src/theory/quantifiers/sygus_process_conj.cpp

```diff
@@ -27,11 +27,546 @@ namespace CVC4 {
+        }
+      }
+    }
+    else
+    {
+      if (def.isNull())
+      {
+        if (i != rid)
+        {
+          // marked as relevant, but template can be set equal to master
+          d_arg_props[i].d_template = d_arg_vars[rid];
+          Trace("sygus-process-arg-deps") << " (new definition, map to master "
+                                          << d_arg_vars[rid] << ")."
+                                          << std::endl;
+        }
+        else
+        {
+          d_arg_props[i].d_relevant = true;
+          Trace("sygus-process-arg-deps") << " (marked relevant)." << std::endl;
+        }
+      }
+      else
+      {
+        // has new definition
+        d_arg_props[i].d_template = def;
+        Trace("sygus-process-arg-deps") << " (new definition " << def << ")."
+                                        << std::endl;
+      }
+    }
+  }
+  return rid;
+}
+
+void CegConjectureProcessFun::processTerms(
+    std::vector<Node>& ns,
+    std::vector<Node>& ks,
+    Node nf,
+    std::unordered_set<Node, NodeHashFunction>& synth_fv,
+    std::unordered_map<Node,
+                       std::unordered_set<Node, NodeHashFunction>,
+                       NodeHashFunction>& free_vars)
+{
+  Assert(ns.size() == ks.size());
+  Trace("sygus-process-arg-deps") << "Process " << ns.size()
+                                  << " applications of " << d_synth_fun << "..."
+                                  << std::endl;
+
+  // get the relevant variables
+  // relevant variables are those that appear in the body of the conjunction
+  std::unordered_set<Node, NodeHashFunction> rlv_vars;
+  Assert(free_vars.find(nf) != free_vars.end());
+  rlv_vars = free_vars[nf];
+
+  // get the single occurrence variables
+  // single occurrence variables are those that appear in only one position,
+  // as an argument to the function-to-synthesize.
+  std::unordered_map<Node, bool, NodeHashFunction> single_occ_variables;
+  for (unsigned index = 0; index < ns.size(); index++)
+  {
+    Node n = ns[index];
+    for (unsigned i = 0; i < n.getNumChildren(); i++)
+    {
+      Node nn = n[i];
+      if (nn.isVar())
+      {
+        std::unordered_map<Node, bool, NodeHashFunction>::iterator its =
+            single_occ_variables.find(nn);
+        if (its == single_occ_variables.end())
+        {
+          // only irrelevant variables
+          single_occ_variables[nn] = rlv_vars.find(nn) == rlv_vars.end();
+        }
+        else
+        {
+          single_occ_variables[nn] = false;
+        }
+      }
+      else
+      {
+        std::unordered_map<Node,
+                           std::unordered_set<Node, NodeHashFunction>,
+                           NodeHashFunction>::iterator itf = free_vars.find(nn);
+        Assert(itf != free_vars.end());
+        for (std::unordered_set<Node, NodeHashFunction>::iterator itfv =
+                 itf->second.begin();
+             itfv != itf->second.end();
+             ++itfv)
```

Comment:
This is again a case where a for each loop would be more compact and easier to read.